File: Code Forces/Practice/Graph/bear_and_friendship.py
#	!/usr/bin/env python3
#	coding: UTF-8
#	Modified: <14/Feb/2019 11:25:46 AM>


#	✪ H4WK3yE乡
#	Mohd. Farhan Tahir
#	Indian Institute Of Information Technology (IIIT),Gwalior

#	Question Link
#
#

# ///==========Libraries, Constants and Functions=============///


import sys
import threading
sys.setrecursionlimit(10**6)


inf = float("inf")
mod = 1000000007


def get_array(): return list(map(int, sys.stdin.readline().split()))


def get_ints(): return map(int, sys.stdin.readline().split())


def input(): return sys.stdin.readline()

# ///==========MAIN=============///


N = 150005
graph = [[] for _ in range(N)]
visited = [False] * N
vertices, edges = 0, 0
flag = 1


def explore(node):
    global visited, flag, vertices, edges
    visited[node] = True
    vertices += 1
    edges += len(graph[node])

    for neighbour in graph[node]:
        if visited[neighbour] == False:
            explore(neighbour)


def dfs(n):
    global visited, flag, vertices, edges
    for i in range(1, n + 1):
        if visited[i] == False:
            vertices, edges = 0, 0
            explore(i)
            if edges != vertices * (vertices - 1):
                flag = 0
                return


def main():
    n, m = get_ints()
    for _ in range(m):
        u, v = get_ints()
        graph[u].append(v)
        graph[v].append(u)
    dfs(n)
    print("YES" if flag == 1 else "NO")


if __name__ == "__main__":
    threading.stack_size(10**8)
    t = threading.Thread(target=main)
    t.start()
    t.join()

File: Code Forces/Practice/Graph/bear_and_friendship.cpp
/*
                C++
                encoding:UTF-8
                Modified: <14/Feb/2019 11:17:47 AM>

                ✪ H4WK3yE乡
                Mohd. Farhan Tahir
                Indian Institute Of Information Technology (IIIT),Gwalior
*/

#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define mod 1000000007
#define sum(v) accumulate(v.begin(), v.end(), 0);
#define reset(a, b) memset(a, b, sizeof(a));
#define pb push_back

const int N = 150005;
vector<int> graph[N];
int visited[N];
int vertices, edges;
int flag = 1;

void explore(int node) {
  visited[node] = 1;
  vertices++;
  edges += graph[node].size();
  for (int neighbour : graph[node]) {
    if (!visited[neighbour]) {
      explore(neighbour);
    }
  }
}

void dfs(int n) {
  for (int i = 1; i <= n; i++) {
    if (!visited[i]) {
      vertices = 0, edges = 0;
      explore(i);
      if (edges != (ll)vertices * (vertices - 1)) {
        flag = 0;
        return;
      }
    }
  }
}

int main() {
  ios_base::sync_with_stdio(false), cin.tie(0), cout.tie(0);
  int n, m;
  cin >> n >> m;
  int u, v;

  for (int i = 0; i < m; i++) {
    cin >> u >> v;
    graph[u].pb(v);
    graph[v].pb(u);
  }
  dfs(n);
  cout << (flag ? "YES" : "NO") << "\n";
  return 0;
}
